File: 21-1.rb
#!/usr/bin/env ruby

class DeterministicDie
  attr_accessor :num_rolls
  attr_accessor :state

  def initialize
    @state = 0
    @num_rolls = 0
  end

  def roll
    @num_rolls += 1
    new_state = @state + 1
    new_state = 1 if new_state == 101
    @state = new_state
  end

  def to_s
    "<#{self.class}: #{@state}>"
  end

  def inspect
    to_s
  end
end

class Game
  def initialize(die, p1, p2)
    @die = die
    @scores = [0, 0]
    @positions = [p1 - 1, p2 - 1]
  end

  def move!(player, spaces)
    @positions[player] = (@positions[player] + spaces) % 10
  end

  def go(player)
    move = 0
    3.times { |_| move += @die.roll }
    new_position = move!(player, move)
    @scores[player] += new_position + 1
    @positions[player] = new_position
  end

  def round
    go(0)
    return 0 if @scores[0] >= 1000
    go(1)
    return 1 if @scores[1] >= 1000
    false
  end

  def play
    winner = false
    loop do
      winner = round
      break if winner
    end

    return @die.num_rolls * @scores[1 - winner]
  end
end

format = /^Player (?<player>[[:digit:]]) starting position: (?<pos>[[:digit:]]+)$/

lines = File.read('21.input').lines.map(&:strip)

players = [0, 0]
2.times do
  m = format.match(lines.shift)
  players[m['player'].to_i - 1] = m['pos'].to_i
end

die = DeterministicDie.new

game = Game.new(die, players[0], players[1])

print game.play, "\n"

File: 21-1.py
#!/usr/bin/env python

import re

class DeterministicDie:
    def __init__(self):
        self.state = 0
        self.num_rolls = 0

    def roll(self):
        self.num_rolls += 1
        new_state = self.state + 1
        if new_state == 101:
            new_state = 1
        self.state = new_state
        return self.state


class Game:
    def __init__(self, die, p1, p2):
        self.die = die
        self.scores = [0, 0]
        self.positions = [p1 - 1, p2 - 1]

    def move(self, player, spaces):
        self.positions[player] = (self.positions[player] + spaces) % 10
        return self.positions[player]

    def go(self, player):
        move = 0
        for _ in range(3):
            move += self.die.roll()
        new_position = self.move(player, move)
        self.scores[player] += new_position + 1
        self.positions[player] = new_position

    def round(self):
        self.go(0)
        if self.scores[0] >= 1000:
            return 0
        self.go(1)
        if self.scores[1] >= 1000:
            return 1
        return None

    def play(self):
        winner = None
        while True:
            winner = self.round()
            if winner is not None:
                break

        return self.die.num_rolls * self.scores[1 - winner]


format = re.compile("^Player (?P<player>[0-9]) starting position: (?P<pos>[0-9]+)$")

lines = [line.strip() for line in open('21.input').readlines()]

players = [0, 0]
for l in range(2):
  m = format.match(lines[l])
  players[int(m['player']) - 1] = int(m['pos'])

die = DeterministicDie()

game = Game(die, players[0], players[1])

print(game.play())
